lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+			return errors.Wrapf(err, "RemoteNode.HandleVersionMessage: Problem sending version message to peer (id= %d)", rn.id)
+		}
+	}
+
+	// After sending and receiving a compatible version, send the verack message. Notice that we don't wait for the
+	// peer's verack message even if it is an inbound peer. Instead, we just send the verack message right away.
+
+	// Set the latest time by which we should receive a verack message from the peer.
+	rn.verackTimeExpected = time.Now().Add(rn.params.VersionNegotiationTimeout)
+	if err := rn.sendVerack(); err != nil {
+		return errors.Wrapf(err, "RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d)", rn.id)
+	}
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
+	return nil
+}
+
+// sendVerack constructs and sends a verack message to the peer.
+func (rn *RemoteNode) sendVerack() error {
+	verackMsg, err := rn.newVerackMessage()
+	if err != nil {
+		return err
+	}
+
+	if err := rn.SendMessage(verackMsg); err != nil {
+		return errors.Wrapf(err, "RemoteNode.SendVerack: Problem sending verack message to peer (id= %d): %v", rn.id, err)
+	}
+	return nil
+}
+
+// newVerackMessage constructs a verack message to be sent to the peer.
+func (rn *RemoteNode) newVerackMessage() (*MsgDeSoVerack, error) {
+	verack := NewMessage(MsgTypeVerack).(*MsgDeSoVerack)
+	vMeta := rn.getHandshakeMetadata()
+
+	switch vMeta.negotiatedProtocolVersion {
+	case ProtocolVersion0, ProtocolVersion1:
+		// For protocol versions 0 and 1, we just send back the nonce we received from the peer in the version message.
+		verack.Version = VerackVersion0
+		verack.NonceReceived = vMeta.versionNonceReceived
+	case ProtocolVersion2:
+		// For protocol version 2, we need to send the nonce we received from the peer in their version message.
+		// We also need to send our own nonce, which we generate for our version message. In addition, we need to
+		// send a current timestamp (in microseconds). We then sign the tuple of (nonceReceived, nonceSent, tstampMicro)
+		// using our validator BLS key, and send the signature along with our public key.
+		var err error
+		verack.Version = VerackVersion1
+		verack.NonceReceived = vMeta.versionNonceReceived
+		verack.NonceSent = vMeta.versionNonceSent
+		tstampMicro := uint64(time.Now().UnixMicro())
+		verack.TstampMicro = tstampMicro
+		verack.PublicKey = rn.keystore.GetSigner().GetPublicKey()
+		verack.Signature, err = rn.keystore.GetSigner().SignPoSValidatorHandshake(verack.NonceSent, verack.NonceReceived, tstampMicro)
+		if err != nil {
+			return nil, fmt.Errorf("RemoteNode.newVerackMessage: Problem signing verack message: %v", err)
+		}
+	}
+	return verack, nil
+}
+
+// HandleVerackMessage handles a verack message received from the peer.
+func (rn *RemoteNode) HandleVerackMessage(vrkMsg *MsgDeSoVerack) error {
+	if rn.connectionStatus != RemoteNodeStatus_Connected {
```

Comment:
Similar question as above. Make sure the handshake is safe if the peer sends does any of the following:
1. Sends a verack message before a version message
2. Sends a verack message before we send our verack message